use low-level firmware for string descriptors

tlyu opened the pull request:
Use the low-level firmware for sending string descriptors.

This required a change to the low-level API to make the string descriptor count configurable.

tlyu commented:
This is another place where there's probably a bug in the low-level firmware. The enumeration code uses the wrong constant to bounds-check the string index in the GetDescriptor request. The firmware example apps have a `usbd_conf.h` that sets `USB_STRING_COUNT`, but nothing in the firmware reads it.

obra commented:
Is there a way to do this without the change to the low-level firmware? If there isn't, can you throw in a bunch of comments so that someone helpfully upgrading the low-level firmware in the future doesn't accidentally break things?

tlyu commented:
> Is there a way to do this without the change to the low-level firmware?

Yes, but it involves either exposing a latent read overrun vulnerability in the vendor firmware, or padding the array with excess valid string descriptors to work around the overflow. The latter might be feasible with the previous vendor firmware, but the current vendor firmware sets `STR_IDX_MAX`, the bounds checking value, to `0xEFU`, which makes the workaround more unwieldy. (The large value is presumably to support proprietary Microsoft string descriptors (index `0xEE`) in the future, but the (buggy) bounds checking wasn't adjusted to accommodate that.)

The intended API, as evidenced by the example code, appears to be for the application code to set `USB_STRING_COUNT` in `usbd_conf.h` to the maximum string descriptor index, plus one, but that's not the value used in the bounds check. (It seems to be an intended change that wasn't fully implemented for some reason.)

> If there isn't, can you throw in a bunch of comments so that someone helpfully upgrading the low-level firmware in the future doesn't accidentally break things?

I can do that. I should probably split the bugfix to the vendor firmware into its own commit, too.

obra commented:
Ok. It's probably worth pushing what we can upstream.

On Wed, Jan 25, 2023 at 2:51 PM Taylor Yu ***@***.***> wrote:

> Is there a way to do this without the change to the low-level firmware?
>
> Yes, but it involves either exposing a latent read overrun vulnerability
> in the vendor firmware, or padding the array with excess valid string
> descriptors to work around the overflow. The latter might be feasible with
> the previous vendor firmware, but the current vendor firmware sets
> STR_IDX_MAX, the bounds checking value, to 0xEFU, which makes the
> workaround more unwieldy. (The large value is presumably to support
> proprietary Microsoft string descriptors (index 0xEE) in the future, but
> the (buggy) bounds checking wasn't adjusted to accommodate that.)
>
> The intended API, as evidenced by the example code, appears to be for the
> application code to set USB_STRING_COUNT in usbd_conf.h to the maximum
> string descriptor index, plus one, but that's not the value used in the
> bounds check. (It seems to be an intended change that wasn't fully
> implemented for some reason.)
>
> If there isn't, can you throw in a bunch of comments so that someone
> helpfully upgrading the low-level firmware in the future doesn't
> accidentally break things?
>
> I can do that. I should probably split the bugfix to the vendor firmware
> into its own commit, too.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/keyboardio/ArduinoCore-GD32-Keyboardio/pull/19#issuecomment-1404327976>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAALC2CKWUVNJ6IRP5TKGWDWUGU7DANCNFSM6AAAAAASLU6LOQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

tlyu committed:
USBD low-level fix for string desc index

Use USB_STRING_COUNT instead of STR_IDX_MAX to bounds-check the
requests for string descriptors. This seems to be the intent of
the API, because the vendor examples set USB_STRING_COUNT in the
application-specific file usbd_conf.h.

In contrast, STR_IDX_MAX is hard-coded in the low-level firmware's
usbd_enum.h, and might be too high for some applications.

Signed-off-by: Taylor Yu <code@argon.blue>

tlyu committed:
use low-level firmware for string descriptors

Use the low-level firmware for sending string descriptors.

This required a change to the low-level API to make the string
descriptor count configurable.

Signed-off-by: Taylor Yu <code@argon.blue>

obra commented:
I think the only thing missing now is the "bunch of comments so that someone helpfully upgrading the low-level firmware in the future doesn't accidentally break things?" I was looking for in the vendor firmware patch.

obra commented:
(Specifically, I'm looking for the 'why' in the comments, not just the 'what')

tlyu commented:
> I think the only thing missing now is the "bunch of comments so that someone helpfully upgrading the low-level firmware in the future doesn't accidentally break things?" I was looking for in the vendor firmware patch.

Oh, so you would like added source code comments in the patch to the vendor firmware that explicitly say that it's a bugfix and why? So it pops out if someone is diffing against the vendor stock firmware?

obra commented:
Right! Because I've seen this pattern bite folks too many times.

On Mon, Jan 30, 2023 at 5:27 PM Taylor Yu ***@***.***> wrote:

> I think the only thing missing now is the "bunch of comments so that
> someone helpfully upgrading the low-level firmware in the future doesn't
> accidentally break things?" I was looking for in the vendor firmware patch.
>
> Oh, so you would like added source code comments in the patch to the
> vendor firmware that explicitly say that it's a bugfix and why? So it pops
> out if someone is diffing against the vendor stock firmware?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/keyboardio/ArduinoCore-GD32-Keyboardio/pull/19#issuecomment-1409611135>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAALC2F4PUQLGAIZQNYKXKTWVBTBBANCNFSM6AAAAAASLU6LOQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>
